File: java-compat/test/src/main/java/test/Ice/dispatcher/Dispatcher.java
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package test.Ice.dispatcher;

public class Dispatcher implements Runnable, Ice.Dispatcher
{
    private static void
    test(boolean b)
    {
        if(!b)
        {
            throw new RuntimeException();
        }
    }

    public Dispatcher()
    {
        _thread = new Thread(this);
        _thread.start();
    }

    @Override
    public void 
    run()
    {
        while(true)
        {
            Runnable call = null;
            synchronized(this)
            {
                if(!_terminated && _calls.isEmpty())
                {
                    try
                    {
                        wait();
                    }
                    catch(java.lang.InterruptedException ex)
                    {
                    }
                }
                
                if(!_calls.isEmpty())
                {
                    call = _calls.poll();
                }
                else if(_terminated)
                {
                    // Terminate only once all calls are dispatched.
                    return;
                }
            }
            
            if(call != null)
            {
                try
                {
                    call.run();
                }
                catch(Exception ex)
                {
                    // Exceptions should never propagate here.
                    test(false);
                }
            }
        }
    }
    
    @Override
    synchronized public void
    dispatch(Runnable call, Ice.Connection con)
    {
        boolean added = _calls.offer(call);
        assert(added);
        if(_calls.size() == 1)
        {
            notify();
        }
    }

    public void
    terminate()
    {
        synchronized(this)
        {
            _terminated = true;
            notify();
        }
        while(true)
        {
            try
            {
                _thread.join();
                break;
            }
            catch(java.lang.InterruptedException ex)
            {
            }
        }
    }
    
    public boolean
    isDispatcherThread()
    {
        return Thread.currentThread() == _thread;
    }

    private java.util.Queue<Runnable> _calls = new java.util.LinkedList<Runnable>();
    private Thread _thread;
    private boolean _terminated = false;
}

File: python/test/Ice/executor/Dispatcher.py
#!/usr/bin/env python3

# Copyright (c) ZeroC, Inc.

import threading


def test(b):
    if not b:
        raise RuntimeError("test assertion failed")


class Dispatcher:
    def __init__(self):
        self._calls = []
        self._terminated = False
        self._cond = threading.Condition()
        self._thread = threading.Thread(target=self.run)
        self._thread.start()
        Dispatcher._instance = self

    def dispatch(self, call, con):
        with self._cond:
            self._calls.append(call)
            if len(self._calls) == 1:
                self._cond.notify()

    def run(self):
        while True:
            call = None
            with self._cond:
                while not self._terminated and len(self._calls) == 0:
                    self._cond.wait()
                if len(self._calls) > 0:
                    call = self._calls.pop(0)
                elif self._terminated:
                    # Terminate only once all calls are dispatched.
                    return

            if call:
                try:
                    call()
                except Exception:
                    # Exceptions should never propagate here.
                    test(False)

    @staticmethod
    def terminate():
        with Dispatcher._instance._cond:
            Dispatcher._instance._terminated = True
            Dispatcher._instance._cond.notify()

        Dispatcher._instance._thread.join()
        Dispatcher._instance = None

    @staticmethod
    def isDispatcherThread():
        return threading.current_thread() == Dispatcher._instance._thread

    @staticmethod
    def instance():
        return Dispatcher._instance
